### DISCUSSION TITLE
Is there any way to build call graph path?

### DISCUSSION THREAD

**Discussion by ox1234** [github/codeql#7531] [Q&A] [answered]
The codeql document only shows how to generate path query through DataFlow problem.Now i just want to generate call graph path from source to sink and don't want to do any data flow track.Is there a way to do so?

**Comment by smowton** [comment]
So you've got a `DataFlow::PathNode source, DataFlow::PathNode sink` but you want something different to the usual dataflow path? Could you given an example of what you do want?

**Reply by ox1234** [reply]
Now i have one source method and one sink method. I want to know the call graph path from source to sink such as: source() -> func1() -> func2() -> sink(). I don't want to do any data flow tracking. All i just want to get is the call method path from source to sink. So is there any way to do that?

**Comment by intrigus-lgtm** [comment, marked as answer]
Have a look at this discussion: https://github.com/github/codeql/discussions/5353#discussioncomment-439461
Also look at the additional comments from `@Marcono1234`.

I *think* this is what you want :)


(Quote from the above discussion also copied and pasted below)
> Yes, this is possible!
> 
> The site you linked to mentions it [here](https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/#defining-your-own-edges-predicate) and [here](https://codeql.github.com/docs/writing-codeql-queries/creating-path-queries/#generating-path-explanations) although it's easy too miss or easy to underestimate its potential.
> 
> When you use taint or data-flow the `edges` predicate is defined by the `PathGraph` module. But you can also define your own `edges` query-predicate.
> 
> A self-defined query-predicate is used in `@agustingianni`'s [blog post](https://github.blog/2021-02-25-the-little-bug-that-couldnt-securing-openssl/#finding-all-the-bug-variants). It's relatively easy to port the code to "Java CodeQL".
> 
> Here's my code that **only** creates a path for methods itself and not for the (control flow) basic-blocks. [Link to query](https://lgtm.com/query/2959145059839326070/)
> 
> ```ql
> /**
>  * @kind path-problem
>  */
> 
> import java
> 
> class StartMethod extends Method {
>   StartMethod() { getName() = "validateExpression" }
> }
> 
> class TargetMethod extends Method {
>   TargetMethod() { getName() = "findValue" }
> }
> 
> query predicate edges(Method a, Method b) { a.calls(b) }
> 
> from TargetMethod end, StartMethod entryPoint
> where edges+(entryPoint, end)
> select end, entryPoint, end, "Found a path from start to target."
> ```

**Reply by ox1234** [reply]
It is exactly what i want. Thanks a lot for help!

**Reply by Manouchehri** [reply]
Likewise, thank you! This should really be added to somewhere official on the CodeQL docs, it's incredibly useful.

**Reply by MengJingsong** [reply]
incredibly useful
